### ISSUE TITLE
optimize block shared memory allocation

### ISSUE THREAD

**Issue by BenjaminW3** [alpaka-group/alpaka#409]
For each variable in block shared memory, a single allocation is performed.
As far as I can imagine, CUDA compilers have an advantage here, because they could sum up the size of all static block shared memory allocations at compile time.

If a kernel (and the functions called from it) require multiple block shared variables, multiple allocations are exeuted. It would be more optimal to do only one allocation for all those shared variables **(1)**. But we could combine not only allocations within a block but for the whole grid **(2)** or even the whole CPU device **(3)**.
Possible solutions for the optimizations:
* **(1)** we could add a block-shared-memory allocator which uses a preallocated memory pool. We could use MallocMC, but this might be too heavy for this first level of optimization. Alternatively a simple buffer with a hard coded limit from which all allocations are served could be used. If this buffer is fully utilized, the next allocation could either throw an exception (what does CUDA do if the shared memory is full?) or allocate new memory.
* **(2)** same as the above version, but the memory pool is held by the executor and shared by all accelerators / block-shared-memory allocators
* **(3)** same as the above version, but the memory pool is held by the cpu device and shared by all executors / accelerators / block-shared-memory allocators

**Comment by psychocoderHPC** [comment]
This is a patch a of possible solution which use a limited amount of shared memory located in the stack  (CUDA also limits the shared memory per multiprocessor depending of the architecture).
```diff
--- a/thirdParty/alpaka/include/alpaka/block/shared/st/BlockSharedMemStNoSync.hpp
+++ b/thirdParty/alpaka/include/alpaka/block/shared/st/BlockSharedMemStNoSync.hpp
@@ -30,6 +30,7 @@
 
 #include <vector>                           // std::vector
 #include <memory>                           // std::unique_ptr
+#include <array>
 
 namespace alpaka
 {
@@ -50,7 +51,9 @@ namespace alpaka
                     //-----------------------------------------------------------------------------
                     //! Default constructor.
                     //-----------------------------------------------------------------------------
-                    ALPAKA_FN_ACC_NO_CUDA BlockSharedMemStNoSync() = default;
+                    ALPAKA_FN_ACC_NO_CUDA BlockSharedMemStNoSync() : memOffset(0u)
+                    {
+                    }
                     //-----------------------------------------------------------------------------
                     //! Copy constructor.
                     //-----------------------------------------------------------------------------
@@ -73,13 +76,9 @@ namespace alpaka
                     ALPAKA_FN_ACC_NO_CUDA /*virtual*/ ~BlockSharedMemStNoSync() = default;
 
                 public:
-                    // TODO: We should add the size of the (current) allocation.
-                    // This would allow to assert that all parallel function calls request to allocate the same size.
-                    std::vector<
-                        std::unique_ptr<
-                            uint8_t,
-                            boost::alignment::aligned_delete>> mutable
-                        m_sharedAllocs;
+
+                    std::array<uint8_t, 128 * 1024> mutable m_sharedAllocs;
+                    size_t mutable memOffset;
                 };
 
                 namespace traits
@@ -102,13 +101,12 @@ namespace alpaka
                             block::shared::st::BlockSharedMemStNoSync const & blockSharedMemSt)
                         -> T &
                         {
-                            blockSharedMemSt.m_sharedAllocs.emplace_back(
-                                reinterpret_cast<uint8_t *>(
-                                    boost::alignment::aligned_alloc(core::vectorization::defaultAlignment, sizeof(T))));
+                            uint8_t* mem = blockSharedMemSt.m_sharedAllocs.data() + blockSharedMemSt.memOffset;
+                            blockSharedMemSt.memOffset += sizeof(T);
                             return
                                 std::ref(
                                     *reinterpret_cast<T*>(
-                                        blockSharedMemSt.m_sharedAllocs.back().get()));
+                                        mem));
                         }
                     };
                     //#############################################################################
@@ -125,7 +123,7 @@ namespace alpaka
                             block::shared::st::BlockSharedMemStNoSync const & blockSharedMemSt)
                         -> void
                         {
-                            blockSharedMemSt.m_sharedAllocs.clear();
+                            blockSharedMemSt.memOffset = 0u;
                         }
                     };
                 }

```

**Comment by psychocoderHPC** [comment]
Shareing one allocation for all blocks is not good, this will result in cache misses because the memory than is not private on the used core.

**Comment by BenjaminW3** [comment]
As long as every two allocations are not in the same cache line, they would not have false sharing problems. Serving the allocations with a distance of a cache line would waste some memory and the cache line size is system dependent but this should be much faster.
For example all modern Xeon processors (also the Phi) have 64 bytes cache line size. In the worst case of 1 byte allocations we would waste 63 bytes, but with high precision 8 byte values we would waste 'only' 56 bytes (87,5%) . However, cache size should not be a problem on those processors.

**Comment by j-stephan** [comment]
This was mostly solved by #1033. The one thing left to do is the increase of the default shared memory size in `CMakeLists.txt`. @psychocoderHPC will open a separate PR for this.
